src/docfx/lib/watch/Watcher.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+
+namespace Microsoft.Docs.Build
+{
+    public static class Watcher
+    {
+        private static readonly AsyncLocal<ImmutableStack<IFunction>> t_callstack = new AsyncLocal<ImmutableStack<IFunction>>();
```

Comment:
>ImmutableStack [](start = 43, length = 14)

question: can we replace the `ImmutableStack` with a `Stack`?

Reply:
No, in case of a parallel call, each thread get the same `Stack` and causes thread safety problem. `ImmutableStack` forks a local copy of the stack on each thread.